### ISSUE TITLE
Metrics data granularity is not respected

### ISSUE THREAD

**Issue by tomkerkhove** [tomkerkhove/promitor#198]
Promitor is currently scraping metrics from Azure Monitor with an aggregation of 5 minutes which is hard-coded.

As part of #60 the plan was to align this with the scraping CRON of Promitor itself but this is not ideal as this is enforced for every metric which can result in data loss or stale data.

This should be defined how to proceed and make required changes to fix this.

<details>
<summary><i>Original concerns reported by @jdevalk2</i></summary>
I have an observation, which I think is an issue, related to this one, but not sure if its its own issue.

Sorry for the long story in advance.
The short of it is: When retrieving metrics, then Promitor should respect the underlying metric data granularities, and the data granularity should (probably?) be configurable on the Generic Metric definition configuration.

Here goes :

First consider a timeline, Azure Monitor will log a datapoint at intervals.

Then on top of this there is the concept of Azure Monitor data granularity.  Which data granularities are supported is depending on the underlying resource. When getting the resource metric definition, it will list its supported data granularity.  Common ones are 30s, 1m, 15m, 30m, 1h, 6h, 12h, 1d

Azure monitor already does aggregation from "raw" datapoints into the above data granularities.
Lets say for example the data granularity is one minute. In that minute AM could have retrieved 2 "raw" datapoints from the resource, those will then be aggregated (sum, total, avg, min, max)

Raw: 10 bytes at T + 15 seconds
Raw: 25 bytes at T + 20 seconds

Then 1 minute data granularity interval for TOTAL aggregation could be 35 bytes.
If those points are the only points in 15 minutes then
15 minute data granularity for TOTAL aggregation is 35 bytes.
1 hour data granularity for TOTAL aggregation is 35 bytes.
etc.

Now we have a cron job which fires data collector process every X interval.
Data collector process purpose is to get the latest measured datapoint for the observed metric.

It works by every interval the past 5 days of data is grabbed, in metric "data granularity" intervals.
That data granularity currently is set to 5 minutes, hardcoded.  

There is a TODO in the code that mentions that should be synced with the CRON job interval. I think though these are both related to time but they are different and should therefore not be synced.

First consideration why I think this wouldn't be ideal is you could have many metrics configured. Each with a different data granularity so in that case it would be hard to have 1 CRON job schedule that would match the data granularity of all underlying metric configured to be measured.

Second consideration like said I think the data granularity instead of being 5m hardcoded should be configurable by the end user in the metric configuration.
In the Azure Monitor query API interval dictates the data granularity series (aggregated by Azure Monitor internally) that is returned.

Third consideration with the granularity hardcoded to 5m, if interested in lower granularity data say per minute, and Prometheus is set to scrape every minute, then 5 minutes is the best resolution you are ever going to see on the metric. And when the metric is updated it might be a couple of minutes off the latest value.

Below I did a quick example of what I mean (I am not that proficient in tables so it looks ugly but hopefully it's clear enough :-))

This is the desired behaviour I think:

If the scraper runs after T+7 minutes you would **expect V7 value for metric1 and V2 value for metric2**

```
Time   Metric1   Metric 2   Scraper
T+0     V1             V1
T+1     V2
T+2     V3 
T+3     V4
T+4     V5             V2  < latest
T+5     V6
T+6     V7 < latest                
T+7     V8                         << Scraper process run here
```

In the current build you would have
If the scraper runs after T+7 minutes again with 5 minute hardcoded data granularity
You would get **V5 value for metric 1 and V2 value for metric2**

```
Time   Metric1   Metric 2   Scraper
T+0     V1             V1
T+1     V2
T+2     V3 
T+3     V4
T+4     V5  <this one V2  < latest
T+5     V6
T+6     V7               
T+7     V8                         << Scraper process run here
```

*More details in #60*
</details>

**Comment by jdevalk2** [comment]
I wanted to document some findings so any solution could keep this into account.

I noticed why the original Promitor retrieval window was set to 5 minute increments. Taking 1 minute increments, Azure Monitor will return the most current points as 0. 

This would give higher likelihood of getting a value, though at the cost of possibly averaging out missing values (assumably as I do not know how smart Azure Monitor is. It does return 0's for values it doesn't have which does not help. More helpful would have been NULL's)

There seems to be a timelag / data latence for a metric to make its way through Azure into Azure Monitor (understandable) but I have observed this lag to be 2 to 3 minutes. I would not rule out it can be longer or shorter depending on the resource being polled.

This gives a bit of a problem with determining the latest value. As if I read the most current value from the query results returned I would continuously plot zero's.

I can devise some logic to go back in time until the first non-zero value. Though this would impact a resource that would return a lot of zero's interspersed with some real values. Like occasional reads from a disk every couple of minutes.  Anyway logic can be applied here, hopefully it would not require statefulness or the ability to know what happened to a resource before to make a judgement on the current datapoint.

So with this logic in place I get data points that are 2 minutes into the past. Which is OK as you would only get it slightly slower than Azure Monitor directly.  However the time an event happened and the observed value does not include the data latency. Which means a graph plotted from Promitor is 2 minutes behind the graph from Azure Monitor directly - once Azure Monitor has calculated the value for the timespan.

So we need to tell Prometheus in the metric that it is not "current" but that the data point occurred 2 minutes in the past. This is possible by adding a Timestamp value to the metric.

However none of the .NET Prometheus clients currently have an ability to write this timestamp with the metric. The Java library for one does write the timestamp with the metric. The timestamp allows Prometheus to project the metric into the past within reason. So the Azure Monitor and Promitor graphs would be in sync.

I checked the official .NET Prometheus client and the fork thereof. The fork is what I could determine is used within Promitor through some owin middleware.

**Comment by tomkerkhove** [comment]
Thanks for sharing your observations.

I did some digging as well and I see the following action points to get this all sorted out:
### **Provide capability to configure the metric aggregation interval per metric**
_(requesting more insights via [Azure/azure-libraries-for-net  #495](https://github.com/Azure/azure-libraries-for-net/issues/495))_
1. Extend the configuration so that metric aggregation interval is configurable via `azureMetricConfiguration.aggregationInterval`
2. Extend our startup validation to verify that all configured aggregation intervals are supported by checking [MetricDefinition.MetricAvailabilities](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.management.monitor.models.metricdefinition.metricavailabilities?view=azure-dotnet) after listing all metrics definitions
3. Use configured aggregation interval for every metric that we scrape
4. Provide documentation on how you can configure this and what the acceptable values are, depending on the outcome of the issue _(enum, fixed set of intervals or worst case timespans)_

That said, I don't think we should do any metric value determination based on the outcome. If the metrics result with the configured aggregation interval is 0, then that's what we will report.

By providing more insights on when we did our scraping action _, see below,_ then this should give an indication to the user that something is wrong or they are aggregating too aggresively.

### **Change the way we scrape Azure Monitor**
1. We should schedule a scraping action for every single metric in the configuration individually
2. Deprecate the general CRON that we have now

### **Provide insights in when the metric was measured in the scraping endpoint**
- I was thinking of adding this to the description of the metric as a suffix but your suggestion is better.
- Opened [PrometheusClientNet/Prometheus.Client #18](https://github.com/PrometheusClientNet/Prometheus.Client/issues/18) to see if this is coming soon, otherwise suggest to fall back to description.

So bottom line, we are decoupling the scheduling on a per-metric basis, but leave all the hard work up to Azure Monitor. They provide the capability to do this, we only need to open this up and ensure that the configuration is correct. 

Do you have any concerns with the topics above @jdevalk2?

**Comment by jdevalk2** [comment]
I am in agreement with most. Though I have an additional remark on 

> That said, I don't think we should do any metric value determination based on the outcome. If the metrics result with the configured aggregation interval is 0, then that's what we will report.
> By providing more insights on when we did our scraping action , see below, then this should give an indication to the user that something is wrong or they are aggregating too aggressively.

I do not think there is one configuration that you can set that will always guarantee a return non zeroes.

Zeros seem to be returned by Azure monitor for a time interval that it has not aggregated data over yet. 

For example, I know a metric is usually delayed 2 minutes, we know normally the most current 2 data intervals will be zero, and the last known value will be in the 3rd most recent interval. Theoretically I could configure a non-aggressive scrape that states only scrape metrics that are older than 2 minutes.

BUT what defines this two minutes, I have to assume it is arbitrary and on good days it is 2 minutes. On bad days, maybe 10 minutes.  20 minutes?  Who knows.

So I do  not see how this can be a hard coded setting in Promitor and Promitor would either have to do some fuzzy selection of the last known real data value, or Azure Monitor would have to not return zero entries for the values it has not aggregated yet at all. Or it should return null's, so its more easily distinguishable.

**Comment by jdevalk2** [comment]
I asked the question on the Azure Monitor issues page as well, see https://github.com/Azure/azure-libraries-for-net/issues/499  I'm interested to know if there might be a way around this query behaviour.

**Comment by jdevalk2** [comment]
Did more research.

### Staleness 

It looks like for any configured Azure resource to be scraped by Promitor, Promitor will need to run at least once every 5 minutes so it can expose a sample to Prometheus. This is in relation to staleness check. If a timeseries does not have a sample within 5 minutes it is considered stale and no values are returned by Prometheus.

So I think a single Cron loop for all resources that runs less than every 5 minutes would not be a bad idea (?)

My understanding is when Promitor would scrape by the hour, that metric would be considered stale by Prometheus and the timeseries would not show up in queries. I haven't tested this this is just my assumption on the docs so far.

Presentation on staleness:

https://www.slideshare.net/brianbrazil/staleness-and-isolation-in-prometheus-20-promcon-2017

**Caution** The weird thing is the presentation above mentions 4 * distance between last 2 metrics points for the staleness check in Prometheus 2.0 But the 5 minutes can still be found quite pervasive in the Prometheus documentation. 

### Latency and graphing discrepancies

The Azure Monitor latency of 2+ minutes seems like it's nothing that can be done about it. I suppose Promitor scrape resource configuration can be configured with a "hold off" delay to only read Azure Monitor data 2 minutes into the past or older. Nothing newer. Anything newer than 2 minutes runs the risk of being not calculated / present and thus 0.

When graphing such a time series, this does result in an offset in the graph of 2 minutes. Events that show on time T actually happened T - 2 minutes ago. This makes it harder to correlate with other logs and events.  In PromQL you can specify "offset" keyword so I am thinking the same config setting Promitor will get for "hold off reading for 2+ minutes" that same time can be given to the offset keyword.

Net result you still have to wait 2 minutes for the results to show up but when they do at least they will be aligned correctly.

**Comment by tomkerkhove** [comment]
> I am in agreement with most. Though I have an additional remark on
> 
> > That said, I don't think we should do any metric value determination based on the outcome. If the metrics result with the configured aggregation interval is 0, then that's what we will report.
> > By providing more insights on when we did our scraping action , see below, then this should give an indication to the user that something is wrong or they are aggregating too aggressively.
> 
> I do not think there is one configuration that you can set that will always guarantee a return non zeroes.
> 
> The risk of plotting a graph with all zeroes (**at least on PT1M intervals** is very real, and it would have no point. If that same time interval were to be queried 2 minutes later it would have a non zero value, other than 0. Thus making 0 data we scrape initially invalid, and any graph thereon kind of useless)
> 
> Zeros seem to be returned by Azure monitor for a time interval that it has not aggregated data over yet.
> 
> For example, I know a metric is usually delayed 2 minutes, we know normally the most current 2 data intervals will be zero, and the last known value will be in the 3rd most recent interval. Theoretically I could configure a non-aggressive scrape that states only scrape metrics that are older than 2 minutes.
> 
> BUT what defines this two minutes, I have to assume it is arbitrary and on good days it is 2 minutes. On bad days, maybe 10 minutes. 20 minutes? Who knows.
> 
> So I do not see how this can be a hard coded setting in Promitor and Promitor would either have to do some fuzzy selection of the last known real data value, or Azure Monitor would have to not return zero entries for the values it has not aggregated yet at all. Or it should return null's, so its more easily distinguishable.

I think you've missed my point in that comment. It's not up to Promitor to decide whether or not the outcome is considered legit. If you configure the aggregation interval to be 1 min and the outcome is `0`, then that's what you'll get.

That's one of the main drivers why we will add a stamp in the scraping endpoint so that you know when it was scraped and it's up to the person whom configured that 1 min interval to increase this. Promitor should not do any actions based on that because it has no idea what is considered a valid metric value. Promitor is just a scraper based on your configuration, not a processor.

## Other topics
- **Staleness** - Honestly, if you configure Promitor to have a granularity of 30 min, we should only check every 30 min. Would you want this to have a representation of the past 30 min, but refreshed every 5 min for example or what is the use-case here?
- **Latency and graphing discrepancies** - Personally I think we should not shift the sampling period ourselves, but we can provide a configuration field similar to `offset` which allows people to configure this if they want. Good suggestion.
- **Timestamp** - Good point, but we could still suffix this to the description of the metric. When people would report that this is an issue we can still provide a feature flag for this.

**Comment by jdevalk2** [comment]
I think we are saying the same thing.
In case of 1 minute intervals where 2-minute latency is visible more, you can set it to aggressively scrape.
If setting the window to less than 2-minutes where Azure Monitor has data available, you will get zeroes mostly.
I agree that is not up to Promitor to fix and it should return 0. It should also be well documented as it might not be obvious to someone familiar with Azure Monitor query semantics.

So setting to less aggressively scrape is one solution.
Setting the interval to 5 minutes is another solution. It depends though if you can do this based on what you are monitoring.
If the aim is to get the most accurate data at the best resolution, making the window 5 minutes might not be desired for the use case.
Between 2 minute and 5 minute latency could be a world of difference. Servers might catch fire in those additional 3 minutes :-)
That difference between 2 and 5 minutes accuracy might also show up in what alert rules you could write on the data.

Ultimately only the user can make the tradeoffs, so Promitor should just expose the dials that users can use to set up a scraping strategy that works for them.


On staleness,
Not that familiar with the Prometheus .NET library internal behaviour.
I am still not sure if Prometheus has 5 minute for staleness or 4*the interval what is documented somewhere for Prometheus 2.
It depends on what happens when we scrape a 30 minute resource.
If Prometheus scapes Promitor endpoint in the meantime, will that 30 minute metric still show up in the exposition HTTP endpoint with its last known value? Or will it drop off that export?
If it stays on that's great and it can just run every 30 minutes. 
If it drops off if it's not refreshed every run then it would need to be refreshed more often than 30 minutes.

On latency,
Maybe in addition to putting the delay in the description, one other way would be to label the metric.
So if the delay is 2 minutes, its possible to label the metric PT2M or something.
Then if the period ever changes you can still distinguish the metrics in the Prometheus Query by label.
Not sure about the complete impact of this though, need to think this through.

**Comment by tomkerkhove** [comment]
> Setting the interval to 5 minutes is another solution. It depends though if you can do this based on what you are monitoring.
If the aim is to get the most accurate data at the best resolution, making the window 5 minutes might not be desired for the use case.
Between 2 minute and 5-minute latency could be a world of difference. Servers might catch fire in those additional 3 minutes :-)
That difference between 2 and 5 minutes accuracy might also show up in what alert rules you could write on the data.

Well frankly, this is not something we can tackle as this is enforced by our dependency so I think they are out of bad link in that respect I think. Maybe we should enforce this and only support a min granularity of 5 min.

Personally, I think we need to be pragmatic on this and not overengineer and start with [this approach](https://github.com/tomkerkhove/promitor/issues/198#issuecomment-432397615) and see how things evolve/our users use Promitor:

**Comment by jdevalk2** [comment]
Fair enough, I will see if I can make 5 minutes work for my case with the resources I am currently scraping.  I do want to point out though that in making the minimum interval 5 minutes on a resource that returns by-the-minute data, it does contradict the earlier point that Promitor be a pass through of the data in Azure Monitor to best extend possible (it should not fudge data to best of its abilities)

Also for high accuracy use cases there is still a fallback to use Azure Monitor datasource directly, while what makes Promitor attractive is to have all metrics in Prometheus and use the single Prometheus alerting engine. But that is an architectural trade off with having 2 alerting engines versus data accuracy.

**Comment by tomkerkhove** [comment]
Well, you could still configure a lower interval than 5-min but then I'm not going to guarantee that you're getting the results you are expecting. I think this should be caveated in the docs that it's possible, but we fully rely on Azure Monitor.

I would see 1 min as the min interval to scrape but only support as of ~2-5min. Would that be ok for you?

This would allow you to use it, but it's possible that you get a 0 back if you're too aggressive and allowing us to be just the pass-through without any logic.

**Comment by jdevalk2** [comment]
That will work, as long as it is a configurable setting. Each application would have to decide for itself what resolution data they want to capture and make the accuracy tradeoffs.

**Comment by tomkerkhove** [comment]
Can you elaborate a bit more on that please? What do you want to be configurable, the metrics interval you mean or? If so, this will certainly be the case!

It will use the configured interval to check with Azure Monitor and return the value **as-is**. On startup, Promitor will validate all configured metric intervals and verify that Azure Monitor supportes them. If not, it will throw validation exception.

**Comment by tomkerkhove** [comment]
So I'm going to create dedicated issues for the following changes:

1. Make scraping interval configurable per metric (5 min as default)
2. Provide capability to configure the metric aggregation interval per metric (5 min as default)
3. Deprecate general scraping interval on the container (dep on #1)

There will not be a restriction on 1, for 2 I'm checking if I can automatically resolve the closest span on the fly, otherwise, I need to look at alternatives such as not allowing smaller periods than 5-min.

Regardless of the potential enforcing of min 5min aggregation windows, does that make sense to you @jdevalk2?

**Comment by tomkerkhove** [comment]
Closing this issue for now.
